apps/studio/src/features/editing-experience/components/RootStateDrawer.tsx

```diff
@@ -25,20 +30,66 @@ export default function RootStateDrawer() {
     setPreviewPageState,
   } = useEditorDrawerContext()
 
-  const onDragEnd = (result: DropResult) => {
-    if (!result.destination) return
+  const { pageId, siteId } = useQueryParse(editPageSchema)
+  const { mutate } = trpc.page.reorderBlock.useMutation({
+    onError: (error, variables) => {
+      // NOTE: rollback to last known good state
+      // @ts-expect-error Our zod validator runs between frontend and backend
+      // and the error type is automatically inferred from the zod validator.
+      // However, the type that we use on `pageState` is the full type
+      // because `Preview` (amongst other things) requires the other properties on the actual schema type
+      setPreviewPageState(variables.blocks)
+      // @ts-expect-error See above
+      setSavedPageState(variables.blocks)
+      toast({
+        title: "Failed to update blocks",
+        description: error.message,
+      })
+    },
+  })
 
-    const updatedBlocks = Array.from(savedPageState)
-    // Remove block at source index
-    const [movedBlock] = updatedBlocks.splice(result.source.index, 1)
-    if (movedBlock) {
-      // Insert at destination index
-      updatedBlocks.splice(result.destination.index, 0, movedBlock)
-    }
+  const toast = useToast({ status: "error" })
 
-    setPreviewPageState(updatedBlocks)
-    setSavedPageState(updatedBlocks)
-  }
+  const onDragEnd = useCallback(
+    (result: DropResult) => {
+      if (!result.destination) return
+
+      const from = result.source.index
+      const to = result.destination.index
+
+      if (
+        from >= savedPageState.length ||
+        to >= savedPageState.length ||
+        from < 0 ||
+        to < 0
+      )
+        return
+
+      // NOTE: We eagerly update their page state here
+      // and if it fails on the backend,
+      // we rollback to what we passed them
+      const updatedBlocks = Array.from(savedPageState)
+      const [movedBlock] = updatedBlocks.splice(from, 1)
+
+      if (!!movedBlock) {
+        updatedBlocks.splice(to, 0, movedBlock)
+        setPreviewPageState(updatedBlocks)
+        setSavedPageState(updatedBlocks)
+      }
+
+      // NOTE: drive an update to the db with the updated index
+      mutate({ pageId, from, to, blocks: savedPageState, siteId })
+    },
```

Comment:
i think `onDragEnd` should only call `mutate` instead of adding the stateful logic inside this function. 

For optimistic updates, we should use the `onMutate` function `react-query` explicitly provides. See [docs](https://tanstack.com/query/v4/docs/framework/react/guides/optimistic-updates).

```suggestion
  const onDragEnd = useCallback(
    (result: DropResult) => {
      if (!result.destination) return

      const from = result.source.index
      const to = result.destination.index

      if (
        from >= savedPageState.length ||
        to >= savedPageState.length ||
        from < 0 ||
        to < 0
      )
        return

      // NOTE: drive an update to the db with the updated index
      mutate({ pageId, from, to, blocks: savedPageState, siteId })
    },
```

Reply:
you're right and i did this somewhere else

Reply:
misunderstood - you're still right!

Reply:
i just tried this - this _doesn't_ quite work. this is because the hook gets fired `onDragEnd`. this results in the series of events below: 

(lets assume `[A, B, C]`)
1. user drags and drops a block (now it should logically be `[B, A, C]`)
2. `onDragEnd` gets called to mutate **but state not updated yet**. hence, the state flips back to `[A, B, C]`
3. `onMutate` gets fired 
4. state updated, now we see `[B, A, C]`

practically, this means that there's a moment (albeit very brief) where the order gets messed up before being rearranged. hence, we'd need to perform the state update in the `onDragEnd` hook itself - we cannot do it outside because it'll cause a re-render, leading to the flicker!

i do agree, though, that this isn't ideal. i've committed my last change and i think we can take a look tomorrow or i'll just drop the commit if this is non-ideal

https://github.com/user-attachments/assets/95df71b6-2513-4c2f-8b49-895b4e97ed8c

Reply:
Why not show a loading state on the Custom blocks section while the mutate is ongoing

Reply:
Wouldn't the flickering still happen? Unless we have loading state in the whole section, then it's very disruptive

Reply:
While a state change is in progress, we technically do not want the user to perform more changes right? If so, shouldn't we block further drag and drops before the previous one fully resolves? Or is that assumption invalid?

Reply:
i think both ways work! i was thinking we should allow updates optimistically (when you start processing the new update, the old update finished alr) because the network would be the bottleneck here, not our db. 

i wanted to show a toast and rollback if changes don't go through - so that in the majority of cases (assumption: network latency >> db latency), your dnd isn't blocked and it's a seamless exp.

i think if we do the block editing, **we should only block drag and drop** rather than having the whole drawer be blocked. 

for consideration, a 15kb payload took 9ms on office wifi and ~2s on fast 3g.

Reply:
@seaerchin revert ba. Do in onDragEnd ba. Cannot avoid the flicker because it's always one render cycle behind. don't know why formsg no flicker, not worth it to debug now.

Reply:
re: offline discussion - we're shifting it back to `onDragEnd`